driver-async/src/main/com/mongodb/async/client/ClientSessionBinding.java

```diff
@@ -155,6 +155,7 @@ private boolean isActiveShardedTxn() {
         private AsyncConnectionSource wrapped;
 
         SessionBindingAsyncConnectionSource(final AsyncConnectionSource wrapped) {
+            ClientSessionBinding.this.retain();
```

Comment:
Let's defer these changes to a separate PR/Jira issue.  The issue that we encounter due to this bug is a symptom of the bug in MongoIterableSubscription, but as I understand it if we fix the latter, we won't encounter this.

Reply:
I will create a new ticket for this.

Reply:
Created ticket [JAVA-3508](https://jira.mongodb.org/browse/JAVA-3508) for this issue.